Release notes for edge-21.5.2 (#6145)
linkerd/linkerd2: CHANGES.md

# Changes

## edge-21.5.2

This edge release updates the proxy-init container to check whether the iptables
rules have already been added, which prevents errors if the proxy-init container

> is restarted. Also, the linkerd cli now has tab completion for kubernetes

Reviewer: ```suggestion
is restarted. Also, the Linkerd CLI now has tab completion for kubernetes
```

> resources, saving you precious keystrokes! Finally, the proxy has been updated

Reviewer: Only `viz stat` command has resource completion. There is a change up for all `viz` commands to have resource completion. After that, the other extensions and core still will not have resource completion. I would change this to mention that `viz stat` has it, and more is on the way.

with several fixes and improvements.

* Added instructions to `build.md` for using a locally built proxy
  (thanks @jroper!)

> * Added support for kubernetes resource aware tab completion

Reviewer: As mentioned above this is not as complete and it sounds right now.

* Updated `proxy-init` to skip configuring firewall if rules exists

> * Fixed viz uninstall to delete all RBAC objects (thanks @aryan9600!)

Reviewer: If you look at the rest of this file we go back and forth on how an extension is called out, but I would go with any of the following: "linkerd-viz uninstall", "Viz uninstall", "`viz uninstall`", etc.

Basically making it obvious this is an extension.

```suggestion
* Fixed Viz uninstall to delete all RBAC objects (thanks @aryan9600!)
```

Author: This is a really good callout because the inconsistency can lead to confusion.

What do you think of:
- using `linkerd-viz` when referring to the extension itself (code, charts, files, architecture, etc.)
- using `viz <command>` when referring to a subcommand that can be run from the CLI?

Reviewer: Yea that seems good to me! I remember discussing this before; it may have been in the original extensions RFC but I'm not positive. Either way this seems like a good way to refer to things.

* Improved diagnostics for rejected profile discovery

> * Added proxy logic to set the l5d-client-id header

Reviewer: ```suggestion
* Added an l5d-client-id header on mutually-authenticated inbound requests so that applications can discover the client's identity.
```

* Reduced proxy resource usage when there are no profiles

> * Updated proxy to fail connections when HTTP detection fails

Reviewer: Hm where is this note from? Is this supposed to be about when `l5d-dst-override` header is present? I would definitely mention include more detail about this note because the proxy doesn't fail when HTTP detection fails.

Author: This note and the one below come from the [notes](https://github.com/linkerd/linkerd2/pull/6107) for proxy v2.143.0

This specific note comes from [linkerd2-proxy/pull/979](https://github.com/linkerd/linkerd2-proxy/pull/979)

Reviewer: Okay so this is specific to the admin server. I would include that level of detail because everywhere else HTTP detection can fail and things will proceed.

> * Changed proxy to assume meshed connections are HTTP/2

Reviewer: Not sure where this note comes from either.. we've used protocol hints in the proxy to assume meshed connections are HTTP/2 so this shouldn't be new.

Author: This note comes from [linkerd2-proxy/pull/982](https://github.com/linkerd/linkerd2-proxy/pull/982)

Reviewer: Okay so again this is specific to the admin server. I think this should just be combined with the other mentioned change about the admin server.

Author: 👍

Author: @kleimkuhler since these two items are specific to the admin server, are they user-facing enough to be included? I'd be fine with removing them if they will have no impact on users

Reviewer: I think they are probably still worth keeping as a bullet even though it is small changes. I think they can be consolidate into:

```
* Changed the admin server to assume all meshed connections are HTTP/2 and fail connections when that is not the case
```

> * Updated the proxy to require the `l5d-dst-override` header

Reviewer: This is very specific to outbound requests for proxy that is in ingress-mode. The header is not required otherwise.

Author: Thanks for clarifying, I updated the note to specify outbound requests in ingress mode

# Changes

## edge-21.5.2

This edge release updates the proxy-init container to check whether the iptables
rules have already been added, which prevents errors if the proxy-init container
is restarted. Also, the `viz stat` command of the linkerd CLI now has tab

> completion for kubernetes resources, saving you precious keystrokes! Finally,

Reviewer 2: ```suggestion
completion for Kubernetes resources, saving you precious keystrokes! Finally,
```

the proxy has been updated with several fixes and improvements.

* Added instructions to `build.md` for using a locally built proxy
  (thanks @jroper!)

> * Added support for kubernetes resource aware tab completion to the `viz stat`

Reviewer 2: ```suggestion
* Added support for Kubernetes resource aware tab completion to the `viz stat`
```

This edge release updates the proxy-init container to check whether the iptables
rules have already been added, which prevents errors if the proxy-init container

> is restarted. Also, the `viz stat` command of the linkerd CLI now has tab

Reviewer: Nit: I think users can assume we're talking about the Linkerd CLI 😬

I also think it's still worth noting that more completions are on the way!

```suggestion
is restarted. Also, the `viz stat` command now has tab
```

completion for kubernetes resources, saving you precious keystrokes! Finally,
the proxy has been updated with several fixes and improvements.

* Added instructions to `build.md` for using a locally built proxy
  (thanks @jroper!)
* Added support for kubernetes resource aware tab completion to the `viz stat`
  command
* Updated `proxy-init` to skip configuring firewall if rules exists
* Fixed viz uninstall to delete all RBAC objects (thanks @aryan9600!)
* Improved diagnostics for rejected profile discovery

> * Added an l5d-client-id header on mutually-authenticated inbound requests so

Reviewer: ```suggestion
* Added the `l5d-client-id` header on mutually-authenticated inbound requests so

  that applications can discover the client's identity.
* Reduced proxy resource usage when there are no profiles
* Updated proxy to fail connections when HTTP detection fails
* Changed proxy to assume meshed connections are HTTP/2
* Updated the proxy to require the `l5d-dst-override` header on outbound

>   requests when the proxy is in `ingress` mode

Reviewer: nit: You say "`ingress` mode" here and "ingress-mode" in the bullet below. How about going with "ingress-mode"?
